Sensors To Show Edit form

joshuaunity opened the pull request:
## Description

This PR introduces a feature that allows the `sensors_to_show` data for an asset to be edited directly from the UI via a form. This enhances the user experience by eliminating the need to modify the JSON directly.

## Look & Feel

![Screenshot from 2024-11-11 13-18-38](https://github.com/user-attachments/assets/47e74822-20b8-43dd-aa4d-9c7683f8ef36)

## How to test

1. Vist your assets page [assets](http://localhost:5000/assets)
2. select any asset from the table
3. Navigate to the `edit asset` modal on the left side of the screen
4. scroll down and click on `edit graph`

## Further Improvements

None

## Related Items

This PR closes #1076 

---

- [x] I agree to contribute to the project under Apache 2 License. 
- [x] To the best of my knowledge, the proposed patch is not based on code under GPL or other license that is incompatible with FlexMeasures

Flix6x commented:
The UX is amazing! :heart:

Please also show the entire asset ancestry for each sensor (like in the breadcrumb, including the organisation account) under an **Asset** property (next to **ID** and **Unit**), so they can be more easily distinguished. If it's trivial, let's split off the account. For example,

```
ID: 79869
Unit: kW
Account: Seita
Asset: Some project > Baseline scenario > HQ > PV system
```
Where `PV system` is a child asset of `HQ`, which is a child asset of `Baseline scenarios`, which is a child asset of `Some project`, which is an asset belonging to the `Seita` account.

Hint: use [this method](https://github.com/FlexMeasures/flexmeasures/blob/c7b37366786849e1e5a5e2eb2832662fb68d191e/flexmeasures/data/models/time_series.py#L70) to fetch the breadcrumb contents.

![image](https://github.com/user-attachments/assets/ed554482-3732-4390-aab0-3b3cd5be47cb)

nhoening reviewed (changes requested):
I have an issue with pre-existing data (in the `sensors_to_show` attribute) not being loaded into this UI.

Let me know if you can reproduce that.

nhoening commented:
I'd like the "Edit" button next to the graph title, not below. Ideally no button (clicking the title means editing) but next to it is also fine (under it eats a lot of vertical space)

Flix6x commented:
> I have an issue with pre-existing data (in the `sensors_to_show` attribute) not being loaded into this UI.
> 
> Let me know if you can reproduce that.

Just checking: if `sensors_to_show` is empty, by default the first 2 sensors of the asset are shown. That is not the situation you are encountering, right?

joshuaunity commented on flexmeasures/ui/templates/crud/asset.html:
```diff
@@ -369,6 +453,257 @@ <h3>All child assets for {{ asset.name }}</h3>
+        if (!response.ok) {
+            showToast("Failed to update the asset", "error");
+        } else{
+            document.dispatchEvent(new Event('refreshGraph'));
```
noted

joshuaunity commented:
> I have an issue with pre-existing data (in the `sensors_to_show` attribute) not being loaded into this UI.

Can you give some insight as to how this happened on your end?

joshuaunity commented:
> I have an issue with pre-existing data (in the `sensors_to_show` attribute) not being loaded into this UI.
> 
> Let me know if you can reproduce that.

When you mean not loaded in the UI, do you mean in the edit form? or the graph/dashboard

nhoening commented:
The graph loaded them, but the edit form (the one from this PR) did not.

nhoening reviewed (changes requested):
I tested again and indeed the existing graphs load in the form! :)

Some things I'd like:

- Move the Edit button next to title
- How do I add a new graph?
- In the search, find sensors on the asset and public ones. I believe now it limits to the account of the user.

joshuaunity commented:
> I tested again and indeed the existing graphs load in the form! :)
> 
> Some things I'd like:
> 
> * Move the Edit button next to title
> * How do I add a new graph?
> * In the search, find sensors on the asset and public ones. I believe now it limits to the account of the user.

I've moved the button, you can check it out

To add sensors to the graph you just fetch the sensors by clicking on the `fetch sensors` button  and add the ones you want

Currently, the API fetches all sensors by default. However, we’ve introduced a new feature to filter sensors by asset_id.

Should we modify the API by default to fetch only sensors from the specified asset (using the asset_id filter) and optionally include public sensors (using the include_public_assets filter)?

joshuaunity commented:
> The UX is amazing! ❤️
> 
> Please also show the entire asset ancestry for each sensor (like in the breadcrumb, including the organisation account) under an **Asset** property (next to **ID** and **Unit**), so they can be more easily distinguished. If it's trivial, let's also split off the account. For example,
> 
> ```
> ID: 79869
> Unit: kW
> Account: Seita
> Asset: Some project > Baseline scenario > HQ > PV system
> ```
> 
> Where `PV system` is a child asset of `HQ`, which is a child asset of `Baseline scenarios`, which is a child asset of `Some project`, which is an asset belonging to the `Seita` account.
> 
> Hint: use [this method](https://github.com/FlexMeasures/flexmeasures/blob/c7b37366786849e1e5a5e2eb2832662fb68d191e/flexmeasures/data/models/time_series.py#L70) to fetch the breadcrumb contents.
> 
> ![image](https://private-user-images.githubusercontent.com/30658763/380683530-ed554482-3732-4390-aab0-3b3cd5be47cb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAxNDg1MzgsIm5iZiI6MTczMDE0ODIzOCwicGF0aCI6Ii8zMDY1ODc2My8zODA2ODM1MzAtZWQ1NTQ0ODItMzczMi00MzkwLWFhYjAtM2IzY2Q1YmU0N2NiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEwMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMDI4VDIwNDM1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWFhMjNlZDAwMTI3MzNjODZlNzhiMTUzNmNlMGE0NjExNTRlNTdkYzNlNmZlMWI3ODhjOTRiYTVmMDVlZGM4NTUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.fv2cMeaqeE7X7bclMLv4PR9AMdUZJGiOph2Pee6wQ9E)

ok, so you suggest the fetch sensor card should show these instead right

nhoening commented:
> To add sensors to the graph

No I meant to add a new graph - I did not see how to do that, maybe I overlooked it

Maybe the API should by default search for sensors on that asset and public ones, not optionally. We can see how well that works ẃhen we use it.

For Felix' request, maybe we want that breadcrumb in the search results *as well* as in the graph cards.

Flix6x commented:
> For Felix' request, maybe we want that breadcrumb in the search results as well as in the graph cards.

Yes, exactly.

nhoening reviewed (changes requested):
Here is my findings from testing (some of these are repetitions, some new):

- I am not sure why the "Add Sensor" button exists. It starts a search on the right, but I can perform the search there. I believe it can go.
- Instead, I am missing a button "Add graph" (maybe left side on the bottom). If I have one graph (see Screenshot), I can add sensors to that, but I don't seem to be able to add another graph.
- The "Edit" Button for the title should go next to the title, not below
- The "fetch Sensors" is not needed if searching happens automatically on every form change. Either not search in every change or get rid of the button I'd say.
- "Account: undefined " should probably say: "Account: Public"
- Sensor search can give the same sensor multiple times, see Screenshot
- I see multiple requests to `null`, see Screenshot
- You also fetch account & asset information from the API for each sensor again, even if you just fetched it for the previous sensor. This should be cached in a dictionary locally.
- Are you already displaying asset information? I don't see any.
- I'd like a spinner to show up when the sensor search in ongoing, so the user knows that something is happening
- I'm searching for "car" and get results that do not seem to have the word "car" in them. Maybe the word matches with something that is not shown? (Can't be the account, as the ones in the second screenshot are public)
- Idea: Sensor IDs can be links to the sensor's page.


![Screenshot from 2024-11-06 10-03-57](https://github.com/user-attachments/assets/9b8bc8be-2e07-459d-99bf-f17a33d9077d)


![Screenshot from 2024-11-06 10-17-52](https://github.com/user-attachments/assets/107636db-35f6-4272-b9cb-565dc47bd329)

joshuaunity commented:
>I'm searching for "car" and get results that do not seem to have the word "car" in them. Maybe the word matches with something that is not shown? (Can't be the account, as the ones in the second screenshot are public)

i cant seem to replicate this particular issue, but if you notice a pattern which always leads to this issue please do let me know.

nhoening reviewed (changes requested):
This improved a lot.
It's a tricky UX problem. I found a few things that would help greatly.
(I did not review the code this time)

- When I open the modal, no graph is selected. However, if I add a sensor, it gets added to one that already exist (probably it was selected, but that was not visible)
- When I added a sensor, the highlighting of the currently selected graph disappears. Also when I delete a sensor. This should remain more stable.
- The above also happens when I perform a new sensor search (the currently selected graph gets deselected). This affects not only visual selection, but also underlying. If I add a sensor, it will be added to a new graph. Actually, I found it hard to add a new sensor which I have to look up first to a currently selected graph. I had to re-select it first, after finding my sensor.
- Clicking on the graph card updates the sensor list, every time. 
Why actually? Maybe it's fine if that does not happen. Now, when I click the title or its "Edit" button, that is not what I want. It also overwrites the search fields I have set. First, let's not update the sensor list if the card is already selected.
- To help user(s) understand better what happens, we could change the text on the button ("Add to Graph") - when a graph is selected, the button text should be "Add to '<TITLE>'  graph". When the selected graph has no title, "Add to selected graph". If no graph is currently selected: "Add to new graph"
- Please add a spinner when the modal is closed and graphs are reloaded.

joshuaunity commented:
Hi, @nhoening this is ready for review now.

Also for the below message, where do you want this spinner?

>Please add a spinner when the modal is closed and graphs are reloaded.

nhoening commented:
> Hi, @nhoening this is ready for review now.
> 
> Also for the below message, where do you want this spinner?
> 
> > Please add a spinner when the modal is closed and graphs are reloaded.

Hm, ideally where the graphs would be.

Flix6x commented on documentation/changelog.rst:
```diff
@@ -17,7 +17,9 @@ New features
 * Speed up loading the accounts page,by making the pagination backend-based and adding support for that in the API [see `PR #1196 <https://github.com/FlexMeasures/flexmeasures/pull/1196>`_]
 * Speed up loading the account detail page by by switching to server-side pagination for assets, replacing client-side pagination [see `PR #1202 <https://github.com/FlexMeasures/flexmeasures/pull/1202>`_]
-* Simplify and Globalize toasts in the flexmeasures project [see `PR #1207 <https://github.com/FlexMeasures/flexmeasures/pull/1207>`_]
+* Simplify and Globalize toasts in the flexmeasures project [see `PR #1207 <https://github.com/FlexMeasures/flexmeasures/pull/1207>_`]
```
This accidentally reverts a fix I made on `main`.

joshuaunity replied:
Oh, you mean the "_" right

Flix6x replied:
Exactly.

nhoening commented:
Text for help icon:

"
Here you can edit what data is shown in the Dashboard Graphs.
Each graph can show one or more sensors (it makes sense to show sensors in a graph that share the same unit).
You can also set the title of each graph and re-order them.
Select a graph (by clicking on its card) to add sensors to it. Sensors can be searched on the right. This will list sensors on the asset or its child asset, as well as public assets.
"

joshuaunity commented:
Hi @nhoening I've worked on the request changes, but there is one I don't really  understand.

>When I click "Add graph", I expect the new card already being visible. Let's add it at the top of the list, and select it right away.

nhoening commented:
> Hi @nhoening I've worked on the request changes, but there is one I don't really  understand.
> 
> >When I click "Add graph", I expect the new card already being visible. Let's add it at the top of the list, and select it right away.

I clicked the button, and no visible change happened - I expected a new card for the graph to appear.

joshuaunity commented:
> > Hi @nhoening I've worked on the request changes, but there is one I don't really  understand.
> > > When I click "Add graph", I expect the new card already being visible. Let's add it at the top of the list, and select it right away.
> 
> I clicked the button, and no visible change happened - I expected a new card for the graph to appear.

Thats weird it works fie on my end

nhoening commented:
[vokoscreenNG-2024-11-18_16-01-44.webm](https://github.com/user-attachments/assets/8584c56b-32e1-4fd9-a7ee-f7909153f782)

This is (after pulling latest code) still what I see.
Which browser are you using?
This is Firefox.

one more idea: When editing the title, it would be great if hitting Enter saved it

joshuaunity commented:
> vokoscreenNG-2024-11-18_16-01-44.webm 
> This is (after pulling latest code) still what I see. Which browser are you using? This is Firefox.
> 
> one more idea: When editing the title, it would be great if hitting Enter saved it

I think I'm more confused. This seems to be behaving as expected, but I think I am missing something.

When "Add Graph" is clicked it fetches the sensors from the API so the user can click on anyone he/she wants to add to the left side.

nhoening commented:
I care less about sensors being already fetched (without specific search parameters). There is a button for that on the right.

But no card on the left is being shown for the new graph in my movie.

joshuaunity commented:
> I care less about sensors being already fetched (without specific search parameters). There is a button for that on the right.
> 
> But no card on the left is being shown for the new graph in my movie.


Im really sorry, but I’m not entirely sure what you mean. Could we have a quick call tomorrow before the standup? maybe that could help me understand what you mean. I also noticed your account is still showing as `undefined`, even though it shouldn’t anymore. A call might help me better understand your environment and scenario.

nhoening commented:
Yes we can talk tomorrow. Let's see if I can make it at 9am.

What I mean is that at the moment you click "Add graph", I don't expect a sensor search to happen on the right. I expect that a new card appears for my graph on the left. Probably the first thing I would do is edit the title, then I would do a sensor search (possibly using my own keyword) to start adding sensors.

- turn automatic sensor search at that event off
- instead, show the graph card

I also see something weird happening with the account, I seem to get the wrong one sometimes.

joshuaunity commented:
> Yes we can talk tomorrow. Let's see if I can make it at 9am.
> 
> What I mean is that at the moment you click "Add graph", I don't expect a sensor search to happen on the right. I expect that a new card appears for my graph on the left. Probably the first thing I would do is edit the title, then I would do a sensor search (possibly using my own keyword) to start adding sensors.
> 
> * turn automatic sensor search at that event off
> * instead, show the graph card
> 
> I also see something weird happening with the account, I seem to get the wrong one sometimes.

Ive pushed an update, you can check it now, also let me know if your accounts are still showing undefined

joshuaunity commented:
>And the issue with Accounts not being shown right is actually a problem with LocalStorage. For instance, I see "Toy Account" on assets when I query sensors who belong to another account, because LocalStorage remembers old values (maybe I was testing/developing with a different database earlier). Btw, you can see LocalStorage in the developer console (F12, in Firefox under "Storage", in Chrome under "Application->Storage").
localStorage.clear(); helps start the page fresh, which is a fix for developers. For now, I believe that is fine, as in production this should not be an issue (LocalStorage is fixed to/within a URL).

Ok, so we leave this as it is?

joshuaunity commented:
>I don't see why there is a sensor search happening with any actions on the left (e.g. it happens when I click "Remove"). When I enter something in the sensor search bar or pick a unit, * then* we should query the sensor API.

I actually tested this on my end again and my previous update already fixed it, below is a video of me replicating the steps you mentioned and there is only one initial call to the sensors API

[Screencast from 2024-11-19 15-21-55.webm](https://github.com/user-attachments/assets/fdb07a4c-2924-40e5-9c34-8b9c5c25b519)

nhoening commented:
> Ok, so we leave this as it is?

Yes

nhoening commented:
In your screen-video I still see a call to the sensor API when you add a graph on the left, at the 4s mark. (I can't see your mouse pointer though, so I am assuming). But it is difficult to tell what happens. I looked at the code for `renderData()` and I don't know why the right side gets shown when we click "Add Graph" and hidden when "Remove" a graph. I would just not do anything with the right side when we change something on the left side. Make them more independent.

Thanks for addressing the smaller items. Here are some things I just found:

- The function `renderData()` (and the variable `dataList`) also have too simple names (and also the comment above the function is not adding anything: "Function to render the data list"). Is this about the data in the graph card?
- Same goes for `filterData()` - be less generic, so the code is more readable (also to yourself in a few weeks :)
- If all that `renderData()` is doing is to render the sensors on the graph card, then I don't think the try/catch blocks you use when you call it make sense. You show a toast saying "Failed to save changes" - but we are not saving any data to the DB right?
- [This text](https://github.com/FlexMeasures/flexmeasures/pull/1212/files#diff-35947be2b92ac9806a287658b5d3040ecf94da73a369c735b2eb701419f03a83R658) ("sensorsContent") should probably not be there?

joshuaunity commented:
>In your screen-video I still see a call to the sensor API when you add a graph on the left, at the 4s mark. (I can't see your mouse pointer though, so I am assuming). But it is difficult to tell what happens. 

I clicked on `Add Graph` when we spoke. You said that when the add graph is clicked, it should add a blank graph and also pre-select that graph. When a graph is selected(it shows the border), it fetches the available sensors so the user can add them instead of having to click on the card manually after "Add graph" to see the available sensors to add.

joshuaunity commented:
>If all that renderData() is doing is to render the sensors on the graph card, then I don't think the try/catch blocks you use when you call it make sense. You show a toast saying "Failed to save changes" - but we are not saving any data to the DB right?

This is correct we initially where, but now we don't anymore, I will refactor that

nhoening commented:
Thanks. I see the appeal but I would suggest to get rid of that for now. It can be confusing and add too many API calls.
If we find later that it would be really useful to add that, we can do that, and then do it due to experience with the UI.

joshuaunity commented:
> Thanks. I see the appeal but I would suggest to get rid of that for now. It can be confusing and add too many API calls. If we find later that it would be really useful to add that, we can do that, and then do it due to experience with the UI.

Ok, so I should seize rendering that part(right part) but leave the border highlight

nhoening commented:
> > Thanks. I see the appeal but I would suggest to get rid of that for now. It can be confusing and add too many API calls. If we find later that it would be really useful to add that, we can do that, and then do it due to experience with the UI.
> 
> Ok, so I should seize rendering that part(the right part) but leave the border highlight.
> 
> Also when I click on another card that should still render the right side right?

Why? I would leave the right side as-is.

joshuaunity commented:
> Thanks. I see the appeal but I would suggest to get rid of that for now. It can be confusing and add too many API calls.
> If we find later that it would be really useful to add that, we can do that, and then do it due to experience with the UI.

Wait, which of the comments was this reply for?

nhoening commented:
About this one, which described automated updates to the sensor list on the rihjt due to changes on the left:

" When a graph is selected(it shows the border), it fetches the available sensors so the user can add them instead of having to click on the card manually after "Add graph" to see the available sensors to add."

nhoening commented on flexmeasures/ui/templates/crud/asset.html:
```diff
@@ -369,6 +483,436 @@ <h3>All child assets for {{ asset.name }}</h3>
+        if (graphIndex !== undefined){
+            savedGraphIndex = graphIndex;
+            selectedGraphTitle = data[graphIndex]?.title;
+            apiSensorsList.innerHTML = ""; 
```
I don't think we need to do this. The sensor search results might be useful, just the graph on the left was not the right one. If I select the right one, I have to repeat the sensor search.

joshuaunity replied:
Thats true, this will be refactored. the expected behavior is to re-render the searched sensors with the currently selected graph index

nhoening commented on flexmeasures/ui/templates/crud/asset.html:
```diff
@@ -369,6 +483,436 @@ <h3>All child assets for {{ asset.name }}</h3>
+    const spinnerElement = document.getElementById('spinnerElement');
+    let rawJSON = "{{ asset.sensors_to_show | safe}}";
+    rawJSON = rawJSON.replace(/'/g, '"');
+    let data = JSON.parse(rawJSON);
```
This needs a more speaking name, for readability and to avoid clashes with other code!

joshuaunity replied:
ill refactor that

nhoening commented on flexmeasures/ui/templates/crud/asset.html:
```diff
@@ -369,6 +483,436 @@ <h3>All child assets for {{ asset.name }}</h3>
+    // Remove sensor from the graph sensor list
+    function removeSensorFromGraph(graphIndex, sensorIndex) {
+        data[graphIndex].sensors.splice(sensorIndex, 1);
+        selectedGraphTitle = data[graphIndex].title;
```
Why do you re-set the graph title here? Seems unnecessary.

joshuaunity replied:
This is actually for the rendered sensors not to show "undefined" cause the graph they were pointing to was removed. But this logic is outdated and there is now a bug with this action.

I have updated the code with a fix to properly handle the state management here

nhoening commented on flexmeasures/ui/templates/crud/asset.html:
```diff
@@ -369,6 +483,436 @@ <h3>All child assets for {{ asset.name }}</h3>
+        const searchInput = event.target.id === "searchInput";
+        const cardBody = event.target.closest(".card-body");
+
+        
```
Please add a comment here (or docstring for the function), explaining your intentions. I see two things happening:

- toggle highlighted card status
- reload sensor list if anything was clicked that is not a card, sensor or search field.

Actually, that last thing is maybe what I would remove. I've said this a few times - removing the sensor search results is not needed unless we re-do the search.
And I'm getting weird behavior, e.g. the sensor results disappear when I am removing a sensor from the graph, probably due to this black-box catch-all `else`.

joshuaunity replied:
I'll add a comment here to better explain whats happening here.

Majorly this is to handle the clicks on a card and it doesn't do any sensor reload, the issue you where facing is probably a bug, but I have done a couple of refactoring so you can try again to see if you are still having that issue.

Also you can see the newly added comment on what this black does.

nhoening reviewed (changes requested):
I like the code!
But I know how I'd love it.

joshuaunity commented on flexmeasures/ui/templates/crud/asset.html:
```diff
@@ -369,6 +483,428 @@ <h3>All child assets for {{ asset.name }}</h3>
+                        <h5 class="card-title pt-2"><b> Sensors: </b></h5>
+                        <div>
+                            ${sensorsContent.length > 0 ? sensorsContent.join("") : "<i>Add sensors on the right side</i>"}
+                            ${multipleUnits ? `<div class="alert alert-warning" role="alert">Note that you are showing sensors with different units</div>` : ""}
```
I think that's true as well

joshuaunity commented on documentation/views/asset-data.rst:
```diff
@@ -25,13 +25,20 @@ The asset page as data dashboard
+^^^^^^^^^^^^^^^^^^^^^^^^^
 
-Finally, it is possible to set custom titles for sensor graphs, by setting within `sensors_to_show` a dictionary with a title and sensor or sensors. For example, ``{"title": "Outdoor Temperature", "sensor": 1}`` or ``{"title": "Energy Demand", "sensors": [2, 3]}`` will display the specified title for the corresponding sensor data.
+The `sensors_to_show`` field controls which sensor data appears in the plot. Use the "Add Graph" button to select from all available sensors associated with the asset, including public sensors, and 
```
Ill, update that

joshuaunity commented on flexmeasures/ui/static/css/flexmeasures.css:
```diff
@@ -1918,3 +1918,16 @@ body.touched [title]:hover:after {
+    display: none !important;
+}
+
+.dropdown-toggle.fa-info::after {
```
Actually for this, if anyone were to use this combination, they would want this effect as it removes the toggle icon(an upsidedown triangle) form the icon it self. If doesn't look pleasant with it included

joshuaunity commented on flexmeasures/ui/templates/crud/asset.html:
```diff
@@ -369,6 +483,436 @@ <h3>All child assets for {{ asset.name }}</h3>
+            spinnerElement.classList.add('hidden-important');
+            apiSensorsListElement.style.display = 'block';
+        } catch (error) {
+            showToast("Failed to filter/search sesnors", "error");
```
>Hm, maybe we should print the error on console.log? 

Ok I can do that and show the toast as well, so the user has a visual that something went wrong

nhoening commented on flexmeasures/ui/templates/crud/asset.html:
```diff
@@ -125,53 +131,60 @@ <h3>Edit {{ asset.name }}</h3>
                                 <div class="form-group form-group-bordered">
-                                    <p>Fields in this section can form part of this asset's <a href="https://flexmeasures.readthedocs.io/en/latest/features/scheduling.html#the-flex-context">flex context</a>. They can also be set in a parent asset or in API requests that trigger schedules.</p>
+                                    <p>Fields in this section can form part of this asset's <a
+                                            href="https://flexmeasures.readthedocs.io/en/latest/features/scheduling.html#the-flex-context">flex
```
With "django" you mean "jinja2"?

Maybe some day (in another PR) we can try using Prettier for this in pre-commit like this:

```
  - repo: https://github.com/pre-commit/mirrors-prettier
    rev: v2.5.1
    hooks:
      - id: prettier
        files: \.(js|css|html|yaml|yml)$
```

This would hopefully install prettier by itself.

joshuaunity replied:
where do i add this

nhoening replied:
This would be added to `.pre-commit-config.yaml`. Maybe give it a try in another branch/PR the we can discuss.

joshuaunity commented on flexmeasures/ui/templates/crud/asset.html:
```diff
@@ -353,6 +374,99 @@ <h3>All child assets for {{ asset.name }}</h3>
+                            </button>
+
+                            <div class="list-container">
+                            <div id="dataList" class="row"></div>
```
ill format this

joshuaunity commented on flexmeasures/ui/templates/crud/asset.html:
```diff
@@ -353,6 +374,99 @@ <h3>All child assets for {{ asset.name }}</h3>
+                            </button>
+
+                            <div class="list-container">
+                            <div id="dataList" class="row"></div>
```
its actually not being used, it part of early development code that's not needed anymore. Ill remove it

joshuaunity commented on flexmeasures/ui/templates/crud/asset.html:
```diff
@@ -369,6 +483,436 @@ <h3>All child assets for {{ asset.name }}</h3>
+    rawJSON = rawJSON.replace(/'/g, '"');
+    let data = JSON.parse(rawJSON);
+    let sensors = [];
+    let editingIndex; 
```
will add some comments to crucial variables

joshuaunity commented on flexmeasures/ui/templates/crud/asset.html:
```diff
@@ -369,6 +483,436 @@ <h3>All child assets for {{ asset.name }}</h3>
 <script src="{{ url_for('flexmeasures_ui.static', filename='js/map-init.js') }}"></script>
 
+<script>
+    let units = {{available_units | tojson}};
```
ok 👍

nhoening commented on flexmeasures/ui/templates/crud/asset.html:
```diff
@@ -184,19 +197,25 @@ <h3>Edit {{ asset.name }}</h3>
+                                    <label for="sensors_to_show" class="col-sm-6 control-label">Sensor Graphs: </label>
+                                
+                                        <button class="btn btn-primary" type="button" data-bs-toggle="modal" data-bs-target="#sensorsToShowForm">
+                                            Edit Graph
```
Actually, "Edit Graphs"

nhoening commented on flexmeasures/ui/templates/crud/asset.html:
```diff
@@ -353,6 +374,99 @@ <h3>All child assets for {{ asset.name }}</h3>
+                                </div>
+                        
+                                <div class="col-4">
+                                    <select class="form-select" aria-label="Default select example"
```
In one of the last commits, you removed  `id="unitsSelect"` from this `select` tag, which stopped the sensor search from working. Adding it back fixed it for me. 
So please add it back :)

joshuaunity replied:
yes, I added it back as I noticed the search wasn't working

nhoening commented on flexmeasures/ui/templates/crud/asset.html:
```diff
@@ -353,6 +374,99 @@ <h3>All child assets for {{ asset.name }}</h3>
+                                </div>
+                        
+                                <div class="col-4">
+                                    <select class="form-select" aria-label="Default select example"
```
I guess "Default select example" does not need to be here?

nhoening commented on flexmeasures/ui/templates/crud/asset.html:
```diff
@@ -369,6 +483,438 @@ <h3>All child assets for {{ asset.name }}</h3>
+                        </div>
+                        <h5 class="card-title pt-2"><b> Sensors: </b></h5>
+                        <div>
+                            ${sensorsContent.length > 0 ? sensorsContent.join("") : "<i>Add sensors on the right side</i>"}
```
As without sensors, the graph is not shown, this should also be rendered in yellow (as the warning about differing units)

nhoening commented on flexmeasures/ui/templates/crud/asset.html:
```diff
@@ -369,6 +483,438 @@ <h3>All child assets for {{ asset.name }}</h3>
+                        
+                        <button class="btn btn-danger btn-sm me-2" onclick="removeGraph(${index})">Remove</button>
+                        <button class="btn btn-secondary btn-sm me-2" onclick="moveGraphUp(${index})" ${index === 0 ? "disabled" : ""}>Move Up</button>
+                        <button class="btn btn-secondary btn-sm" onclick="moveGraphDown(${index})" ${index === data.length - 1 ? "disabled" : ""}>Move Down</button>
```
This does not work perfectly yet.

- I have three graphs.
- The second graph cannot move down, even if it is not last.
- I move the first downwards (to 2nd position), then it has the same problem.

joshuaunity replied:
Could you try this again, may there was a bug introduced in previous commit, cause I am not having this issue on my end

nhoening commented on flexmeasures/ui/templates/crud/asset.html:
```diff
@@ -369,6 +483,438 @@ <h3>All child assets for {{ asset.name }}</h3>
+                    sensorsUnits.push(sensorData.unit);
+            
+                    return `
+                    <div class="p-1 mb-3 border-bottom border-secondary">
```
I find it visually a bit confusing that this "X" icon is top of the `span`. It sits almost within the text, so the second row of text can be mistaken as a different thing, especially if there is only one sensor.

I could improve it by adding it as first element to the `div` and adding `float: right;` - but maybe you have a better idea.

joshuaunity replied:
Ok, i can work on that, there are some Bootstrap classes I could use

nhoening reviewed (changes requested):
This is working pretty sweet now.

Just a few comments, and one serious (but easily fixable) bug.

A question about sensor search: you use pagination, but if the sensor one wants is not on the first page, there is no way to get to the second page.
I don't believe we should add pagination here right now. Maybe we should remove the pagination for now?

Also, there is no way to do a search without entering a keyword or unit. Maybe we could start the dialogue with an initial complete search?

joshuaunity commented on flexmeasures/ui/templates/crud/asset.html:
```diff
@@ -369,6 +483,438 @@ <h3>All child assets for {{ asset.name }}</h3>
+            col.classList.add("col-12", "mb-1");
+
+            const sensorsUnits = [];
+            const newItem = {
```
Yes, we can still create an issue. I could create it.

joshuaunity commented:
> A question about sensor search: you use pagination, but if the sensor one wants is not on the first page, there is no way to get to the second page. I don't believe we should add pagination here right now. Maybe we should remove the pagination for now?

we can remove it for now I think. wth the caching integration it shouldn't affect performance relevantly

> Also, there is no way to do a search without entering a keyword or unit. Maybe we could start the dialogue with an initial complete search?

So do you mean that when the dialogue is opened the right side should be rendered with all the available sensors? or when a card is clicked it should render?

nhoening commented:
Let's try when the dialogue is opened.

joshuaunity commented:
> Let's try when the dialogue is opened.

Ive made the update, you can check it now

nhoening commented:
Now it happens when the page loads, even when the dialogue is not opened. I suggest to only load sensors when the dialogue is opened, with:

```
var initialOpening = False;
$('#myModal').on('shown.bs.modal', function () {
    if initialOpening {
        renderGraphCards();
        filterSensors();
    }
    imitialOpening = false;
});
```

Here, closing and opening again preserves the state.

nhoening commented on flexmeasures/ui/templates/crud/asset.html:
```diff
@@ -347,6 +458,437 @@ <h3>All child assets for {{ asset.name }}</h3>
+    
+                return `
+                    <div class="p-1 mb-3 border-bottom border-secondary">
+                        <div class="d-flex justify-content-between align-items-center">
```
If we don't use `align-items-center` then the "X" will be more top right than just top centered, which I find nicer. Do you see a problem with that?

joshuaunity replied:
No, that's ok. Ill modify it.

nhoening reviewed (changes requested):
- When I change the name of a graph, can we also refresh the sensors on the right, at least the "Add to ... graph" buttons?
- Interestingly, when I move the second graph up to the first position, the former first graph is getting named on these buttons (but is not getting focus necessarily).
- On a related note, when I move a graph up or down, it should also become the focused graph

joshuaunity commented:
>When I change the name of a graph, can we also refresh the sensors on the right, at least the "Add to ... graph" buttons?
Interestingly, when I move the second graph up to the first position, the former first graph is getting named on these buttons (but is not getting focus necessarily).

I have resolved both of these, although it was quite tough figuring out how to manage the unwanted events but I found something, you can test it out now.

>On a related note, when I move a graph up or down, it should also become the focused graph

You mean that it should be the one with the highlighted border. This is something that is working on my end, maybe it was broken before and I didn't notice but testing it right now, it works

joshuaunity commented:
> Now it happens when the page loads, even when the dialogue is not opened. I suggest to only load sensors when the dialogue is opened, with:
> 
> ```
> var initialOpening = False;
> $('#myModal').on('shown.bs.modal', function () {
>     if initialOpening {
>         renderGraphCards();
>         filterSensors();
>     }
>     imitialOpening = false;
> });
> ```
> 
> Here, closing and opening again preserves the state.

just seeing this today but I also implemented it

nhoening commented:
> > When I change the name of a graph, can we also refresh the sensors on the right, at least the "Add to ... graph" buttons?
> > Interestingly, when I move the second graph up to the first position, the former first graph is getting named on these buttons (but is not getting focus necessarily).
> 
> I have resolved both of these, although it was quite tough figuring out how to manage the unwanted events but I found something, you can test it out now.

The first (buttons change on graph name edit) works on Chrome, not on Firefox.
The second seems to be fixed!

> 
> > On a related note, when I move a graph up or down, it should also become the focused graph
> 
> You mean that it should be the one with the highlighted border. This is something that is working on my end, maybe it was broken before and I didn't notice but testing it right now, it works

This does not work for me in either FF or Chrome. But that is not a deal-breaker for me either.

nhoening commented:
> The first (buttons change on graph name edit) works on Chrome, not on Firefox.

I see what it is, actually: IT works when the "Save" button is clicked, but sometimes I hit the Enter key. I see you are catching the Enter event - I could not figure out why the buttons don't update in that case.

joshuaunity commented on flexmeasures/ui/templates/crud/asset.html:
```diff
@@ -347,6 +458,439 @@ <h3>All child assets for {{ asset.name }}</h3>
+                    <div class="d-flex align-items-center">
+                        <div>
+                            ${editingIndex === index
+                            ? `<input type="text" class="form-control mb-2 me-2" id="editTitle_${index}" value="${newItem.title}" onkeydown="handleEnterKeyEvent(event, ${index})" />`
```
that's because i have to catch the correct keypress event first

nhoening replied:
Ah, true.

We could rename the function `handleEnterKeyEventForTitleEditing` as it is only meant for that purpose.

joshuaunity commented:
>This does not work for me in either FF or Chrome. But that is not a deal-breaker for me either.

That's weird, you could send a short video on the experience on your end

joshuaunity commented:
> > > This does not work for me in either FF or Chrome. But that is not a deal-breaker for me either.
> > 
> > 
> > That's weird, you could send a short video on the experience on your end
> 
>  vokoscreenNG-2024-12-03_11-17-27.webm

wow, what if you click on the card itself, it doesn't still highlight

nhoening commented:
> > > > This does not work for me in either FF or Chrome. But that is not a deal-breaker for me either.
> > > 
> > > 
> > > That's weird, you could send a short video on the experience on your end
> > 
> > 
> > vokoscreenNG-2024-12-03_11-17-27.webm
> 
> wow, what if you click on the card itself, it doesn't still highlight

Clicking on the card works well.
Clicking on Edit (title) puts focus (title of buttons on the right changes) but does not draw the blue outline.

joshuaunity commented:
>Clicking on Edit (title) puts focus (title of buttons on the right changes) but does not draw the blue outline.

I revisited this and recalled why it's implemented this way. When a user wants to reorganize multiple graphs without selecting a specific one, having the highlight bar appear would create a cluttered and confusing experience. In such cases, the highlight bar would be all over the place, that's why its not added in the case of directional/organizing buttons clicked

nhoening commented:
Okay, what about clicking Edit selecting the graph (as seen with button names on the right) but the focus border not hapening?

joshuaunity commented:
> Okay, what about clicking Edit selecting the graph (as seen with button names on the right) but the focus border not hapening?

I'm working on that one actually and will push an update

Flix6x commented:
Looks like [our documentation build failed](https://readthedocs.org/projects/flexmeasures/builds/26518393/) just after merging this PR. Probably unrelated, right?

nhoening commented:
> Looks like [our documentation build failed](https://readthedocs.org/projects/flexmeasures/builds/26518393/) just after merging this PR. Probably unrelated, right?

I also noticed (I have turned on email notifications for these failures just now) and tested this a bit this afternoon.
PDF creation is not working, and I could not pinpoint why. It also gows wrong locally, but I didn't seem to see something pointing to this PR. Maybe some dependency version change. Of course I believe we fix dependency versions. Maybe we should turn off PDF docs and make a ticket ...
